### PR TITLE
ANY Aggregator should not skip null values implementation

### PR DESCRIPTION
**Pull request by maytasm3**
Modify ANY aggregator to not skip null values

### Description

Modify ANY aggregator to not skip null values and hence return null values if that is the first value the aggregator encountered. 

- DoubleAnyAggregatorFactory, FloatAnyAggregatorFactory, LongAnyAggregatorFactory instead of extending from SimpleDoubleAggregatorFactory (which skip null values) will instead extend AggregatorFactory and will handles null values by in it's AnyAggregator.
- AnyAggregator will have additional flag to mark if null value is found. If null value is found, all subsequent values will be skipped. 
- AnyAggregator can return null value if null value is he first value the aggregator encountered. 

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by ccaominh** [comment]
Static analysis is flagging unused imports

File: docs/querying/aggregations.md [hunk 1]
```diff
@@ -238,7 +238,7 @@ Note that queries with first/last aggregators on a segment created with rollup e
 
 (Double/Float/Long/String) ANY aggregator cannot be used in ingestion spec, and should only be specified as part of queries.
 
-If `druid.generic.useDefaultValueForNull=true` aggregation can returns the default value for null and does not prefer "non-null" values over the default value for null. If `druid.generic.useDefaultValueForNull=false`, then aggregation will returns any non-null value.
+If `druid.generic.useDefaultValueForNull=true` aggregation can returns the default value for null and does not prefer "non-null" values over the default value for null. If `druid.generic.useDefaultValueForNull=false`, then aggregation will returns any value including null.
```
Comment (suneet-s) [inline docs/querying/aggregations.md]: I think we should re-write this to explain why someone would use this aggregator similar to how it's explained in the snowflake docs - https://docs.snowflake.net/manuals/sql-reference/functions/any_value.html
I'm not sure where the correct place in the docs is to explain this - since technically this is the spec for the native query and we have another page with a spec for sql.

Here's my suggestion for the why:

```
ANY aggregator can be used to simplify and optimize the performance of GROUP BY statements. A common problem for many queries is that the result of a query with a GROUP BY clause can only contain expressions used in the GROUP BY clause itself, or results of aggregate functions

select customer.id , customer.name , sum(orders.value)
    from customer
    join orders on customer.id = orders.customer_id
    group by customer.id , customer.name;

Since we know that customer.id can have only one name, this can be optimized as

select customer.id , ANY(customer.name) , sum(orders.value)
    from customer
    join orders on customer.id = orders.customer_id
    group by customer.id , customer.name;
```

I should also point out, with the current implementation of aggregators, there is no advantage to using an ANY aggregator vs a min aggregator, but maybe that will change in the future 🤷‍♂

Comment (suneet-s) [reply docs/querying/aggregations.md]: Suggested re-wording to explain how null handling works

```
This aggregator will return any value for the provided expression. It does not prefer non-null values over null values. If `druid.generic.useDefaultValueForNull` is set to true, this will not return null, but instead return the default value. Otherwise this aggregator may return null
```

Comment (suneet-s) [reply docs/querying/aggregations.md]: EDIT: @maytasm3 pointed out this benchmark - https://static.imply.io/gianm/vb.html which shows the performance gain of not having to read a value. Since ANY only reads the first value, this will be faster than min, even thought the aggregator has to loop over all values. 

I should have known that... 🤦‍♂

Comment (sthetland) [reply docs/querying/aggregations.md]: +1 to Suneet's rewording. Agreed as well that we should add the "why". Have filed https://implydata.atlassian.net/browse/DOCS-30 to track.

File: docs/querying/sql.md [hunk 2]
```diff
@@ -205,7 +205,7 @@ Only the COUNT aggregation can accept DISTINCT.
...
 |`LATEST(expr, maxBytesPerString)`|Like `LATEST(expr)`, but for strings. The `maxBytesPerString` parameter determines how much aggregation space to allocate per string. Strings longer than this limit will be truncated. This parameter should be set as low as possible, since high values will lead to wasted memory.|
-|`ANY_VALUE(expr)`|Returns any value of `expr`, which must be numeric. If `druid.generic.useDefaultValueForNull=true` this can return the default value for null and does not prefer "non-null" values over the default value for null. If `druid.generic.useDefaultValueForNull=false`, then this will return any non-null value of `expr`|
+|`ANY_VALUE(expr)`|Returns any value of `expr`, which must be numeric. If `druid.generic.useDefaultValueForNull=true` this can return the default value for null and does not prefer "non-null" values over the default value for null. If `druid.generic.useDefaultValueForNull=false`, then this will return any value of `expr` including null|
```
Comment (sthetland) [inline docs/querying/sql.md]: nit, adding a comma: "If `druid.generic.useDefaultValueForNull=true`, this can return..."

Comment (clintropolis) [reply docs/querying/sql.md]: I sort of think this should just be simplified so it is less confusing, maybe something like:

> Returns any value of `expr`, including null.

Also, expr does not need to be numeric since you implemented a string any aggregator in the previous PR.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java [hunk 3]
```diff
@@ -19,44 +19,35 @@
...
 import org.apache.druid.segment.BaseDoubleColumnValueSelector;
 
+import javax.annotation.Nullable;
+
 /**
- * This Aggregator is created by the {@link DoubleAnyAggregatorFactory} which extends from
- * {@link NullableNumericAggregatorFactory}. If null needs to be handle, then {@link NullableNumericAggregatorFactory}
- * will wrap this aggregator in {@link NullableNumericAggregator} and can handle all null in that class.
- * Hence, no null will ever be pass into this aggregator from the valueSelector.
+ * This Aggregator is created by the {@link DoubleAnyAggregatorFactory} which has no special null handling logic.
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java]: I'm not sure this is really worth mentioning. This is the standard case for aggregator implementations, since it is less common for an agg implementation to have magic null handling wrapped around it than just handling the nulls itself.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+ * Base type for buffer based 'any' aggregator for primitive numeric column selectors
+ */
+public abstract class NumericAnyBufferAggregator<TSelector extends BaseNullableColumnValueSelector>
+    implements BufferAggregator
+{
+  private static final byte BYTE_FLAG_IS_NOT_SET = 0;
+  private static final byte BYTE_FLAG_IS_SET = 1;
+  private static final int IS_FOUND_FLAG_OFFSET_POSITION = 0;
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java]: nit: I think this `0` offset variable makes the put and get operations more complicated than they need to be, suggest just dropping this and not adding or removing anything.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+import java.nio.ByteBuffer;
+
+/**
+ * Base type for buffer based 'any' aggregator for primitive numeric column selectors
+ */
+public abstract class NumericAnyBufferAggregator<TSelector extends BaseNullableColumnValueSelector>
+    implements BufferAggregator
+{
+  private static final byte BYTE_FLAG_IS_NOT_SET = 0;
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java]: Please use `NullHandling.IS_NULL_BYTE` and `NullHandling.IS_NOT_NULL_BYTE` to be consistent with other aggregators, at least for the 'is null' byte.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  private static final byte BYTE_FLAG_IS_SET = 1;
+  private static final int IS_FOUND_FLAG_OFFSET_POSITION = 0;
+  private static final int IS_NULL_FLAG_OFFSET_POSITION = IS_FOUND_FLAG_OFFSET_POSITION + Byte.BYTES;
+  private static final int FOUND_VALUE_OFFSET_POSITION = IS_NULL_FLAG_OFFSET_POSITION + Byte.BYTES;
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java]: nit: suggest just making this package private and having subclasses use this directly in their putValue implementations instead of `getFoundValueStoredPosition` function call, and maybe just calling it `VALUE_OFFSET` or `FOUND_VALUE_OFFSET` since the position seems redundant.

File: server/src/main/java/org/apache/druid/server/QueryLifecycle.java [hunk 7]
```diff
@@ -318,14 +318,12 @@ public void emitLogsAndMetrics(
...
       if (e != null) {
         statsMap.put("exception", e.toString());
-
-        if (e instanceof QueryInterruptedException) {
-          // Mimic behavior from QueryResource, where this code was originally taken from.
-          log.noStackTrace().warn(e, "Exception while processing queryId [%s]", baseQuery.getId());
-          statsMap.put("interrupted", true);
-          statsMap.put("reason", e.toString());
-        }
+        // Mimic behavior from QueryResource, where this code was originally taken from.
```
Comment (clintropolis) [inline server/src/main/java/org/apache/druid/server/QueryLifecycle.java]: Why this change? I don't think it is correct, since this stuff should only be set if it was a `QueryInterruptedException`

File: processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java (hunk 4 above)
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java]: Further, since is null is only going to be either 0 or 1, you save a byte per result and could use any of the high bits of this null byte to set whether or not you have found a value. Instead of an offset for null flag, you define a bit mask and your found check becomes something like `buf.get(position) & FOUND_MASK == FOUND_MASK` or whatever.

File: docs/querying/aggregations.md (hunk 1 above)
Comment (maytasm3) [reply docs/querying/aggregations.md]: Added the why (short version of what Suneet wrote)

File: docs/querying/sql.md (hunk 2 above)
Comment (maytasm3) [reply docs/querying/sql.md]: simplified as suggested. 
expr needs to be null for ANY_VALUE(expr)
String column have to be ANY_VALUE(expr, maxBytesPerString)

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java [hunk 8]
```diff
@@ -19,103 +19,214 @@
...
   @Override
   @Nullable
   public Object combine(@Nullable Object lhs, @Nullable Object rhs)
   {
-    if (lhs != null) {
-      return lhs;
-    } else {
-      return rhs;
-    }
+    return lhs;
```
[the remark this answers is not in the document]
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: to be consistent with the new policy of not discriminating null values #equality 
Decided not to make assumption in preferring non-null value.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java [hunk 9]
```diff
@@ -19,103 +19,214 @@
...
     return new CacheKeyBuilder(AggregatorUtil.DOUBLE_ANY_CACHE_TYPE_ID)
         .appendString(fieldName)
-        .appendString(expression)
         .build();
   }
 
+  @Override
+  public String getTypeName()
+  {
+    // if we don't pretend to be a primitive, group by v1 gets sad and doesn't work because no complex type serde
```
[the remark this answers is not in the document]
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: removed.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java (hunk 3 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java]: Removed

File: processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java (hunk 4 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java]: Done

File: processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java (hunk 5 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java]: Done

File: processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java (hunk 6 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java]: Done

File: server/src/main/java/org/apache/druid/server/QueryLifecycle.java (hunk 7 above)
Comment (maytasm3) [reply server/src/main/java/org/apache/druid/server/QueryLifecycle.java]: Changed to just moving the logging out of the if check

Comment (maytasm3) [reply server/src/main/java/org/apache/druid/server/QueryLifecycle.java]: The change is so that we can see the exception (and possibly stacktrace) for exceptions other than QueryInterruptedException too

File: docs/querying/aggregations.md [hunk 10]
```diff
@@ -238,7 +238,7 @@ Note that queries with first/last aggregators on a segment created with rollup e
 
 (Double/Float/Long/String) ANY aggregator cannot be used in ingestion spec, and should only be specified as part of queries.
 
-If `druid.generic.useDefaultValueForNull=true` aggregation can returns the default value for null and does not prefer "non-null" values over the default value for null. If `druid.generic.useDefaultValueForNull=false`, then aggregation will returns any value including null.
+Returns any value including null. This aggregator simplify and optimize the performance by returning the first encoutnered value (including null)
```
Comment (clintropolis) [inline docs/querying/aggregations.md]: typo 'encoutnered' is causing CI failure

File: docs/querying/sql.md [hunk 11]
```diff
@@ -205,7 +205,7 @@ Only the COUNT aggregation can accept DISTINCT.
...
 |`LATEST(expr, maxBytesPerString)`|Like `LATEST(expr)`, but for strings. The `maxBytesPerString` parameter determines how much aggregation space to allocate per string. Strings longer than this limit will be truncated. This parameter should be set as low as possible, since high values will lead to wasted memory.|
-|`ANY_VALUE(expr)`|Returns any value of `expr`, which must be numeric. If `druid.generic.useDefaultValueForNull=true` this can return the default value for null and does not prefer "non-null" values over the default value for null. If `druid.generic.useDefaultValueForNull=false`, then this will return any value of `expr` including null|
+|`ANY_VALUE(expr)`|Returns any value of `expr` including null. `expr` must be numeric. This aggregator simplify and optimize the performance by returning the first encoutnered value (including null)|
```
Comment (clintropolis) [inline docs/querying/sql.md]: nit: should be 'This aggregator _can_ simplify..' 

Should this note also be added to the description of string any value?

File: processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java [hunk 12]
```diff
@@ -32,11 +32,11 @@
...
     implements BufferAggregator
 {
-  private static final byte BYTE_FLAG_IS_NOT_SET = 0;
-  private static final byte BYTE_FLAG_IS_SET = 1;
-  private static final int IS_FOUND_FLAG_OFFSET_POSITION = 0;
-  private static final int IS_NULL_FLAG_OFFSET_POSITION = IS_FOUND_FLAG_OFFSET_POSITION + Byte.BYTES;
-  private static final int FOUND_VALUE_OFFSET_POSITION = IS_NULL_FLAG_OFFSET_POSITION + Byte.BYTES;
+  // Rightmost bit for is null check (0 for is null and 1 for not null)
+  // Second rightmost bit for is found check (0 for not found and 1 for found)
+  private static final byte BYTE_FLAG_FOUND_MASK = 0b0010;
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java]: super super nit, but can you use hex? (`0x02` and `0x01` for `BYTE_FLAG_NULL_MASK`)

File: docs/querying/aggregations.md (hunk 10 above)
Comment (maytasm3) [reply docs/querying/aggregations.md]: oops done.

File: docs/querying/sql.md (hunk 11 above)
Comment (maytasm3) [reply docs/querying/sql.md]: Added 'can'. The string any indicates "Like `ANY_VALUE(expr)`" which implies that whatever is mentioned for the `ANY_VALUE(expr)` also applies to the string any.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java (hunk 12 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/NumericAnyBufferAggregator.java]: Done
